### PR TITLE
Add more APIs in storage sdk

### PR DESCRIPTION
**Pull request by JinmingHu-MSFT**
1. Service Client

- List Containers
- Get User Delegation Key

2. Blob Client

- Set Access Tier
- Copy From Uri
- Abort Copy From Uri
- Create Snapshot
- Undelete

3. Block Blob Client

- Stage Block From Uri

4. Append Blob Client

- Create
- AppendBlock

5. Page Blob Client

- Create
- UploadPages
- UploadPagesFromUri
- ClearPages
- Resize
- GetPageRanges
- StartCopyIncremental

6. Other

- Rewrite xml serializer/deserializer in auto-generated rest client
- Support for conditional headers(If-Modified-Since, If-Unmodified-Since, If-Match, If-None-Match)

### CODE REVIEW

**Comment by katmsft** [comment]
>                 && path[1] == XmlTagName::k_Containers && path[2] == XmlTagName::k_Container)

I think deeply nested `if` statements are more readable than this.

---
Refers to: sdk/storage/inc/blobs/internal/protocol/blob_rest_client.hpp:1212 in 78647d0. [](commit_id = 78647d089befdb8306de1f0f90fac3ac3a1e4d3c, deletion_comment = False)

**Comment by katmsft** [comment]
>         return ret;

This does not except on partial xml body right? Is there a content-length/body length mismatch check then?

---
Refers to: sdk/storage/inc/blobs/internal/protocol/blob_rest_client.hpp:1253 in 78647d0. [](commit_id = 78647d089befdb8306de1f0f90fac3ac3a1e4d3c, deletion_comment = False)

**Comment by JinmingHu-MSFT** [comment]
> > ```
> >             && path[1] == XmlTagName::k_Containers && path[2] == XmlTagName::k_Container)
> > ```
> 
> I think deeply nested `if` statements are more readable than this.
> 
> Refers to: sdk/storage/inc/blobs/internal/protocol/blob_rest_client.hpp:1212 in [78647d0](https://github.com/Azure/azure-sdk-for-cpp/commit/78647d089befdb8306de1f0f90fac3ac3a1e4d3c). [](commit_id = [78647d0](https://github.com/Azure/azure-sdk-for-cpp/commit/78647d089befdb8306de1f0f90fac3ac3a1e4d3c), deletion_comment = False)

Yes, but it's harder to auto-generate

**Comment by katmsft** [comment]
> #include "blob_client_options.hpp"

This file should be named `blob_options.hpp` given that all blob options are in one files now.

---
Refers to: sdk/storage/inc/blobs/blob_client.hpp:6 in 78647d0. [](commit_id = 78647d089befdb8306de1f0f90fac3ac3a1e4d3c, deletion_comment = False)

File: sdk/storage/inc/blobs/internal/protocol/blob_rest_client.hpp [hunk 1]
```diff
@@ -1002,6 +896,304 @@ namespace Azure { namespace Storage { namespace Blobs {
...
+            }
+            else if (std::strcmp(node.Name, "Container") == 0)
+            {
+              path.emplace_back(XmlTagName::k_Container);
+            }
+            else
+            {
+              path.emplace_back(XmlTagName::k_Unknown);
+            }
+            if (path.size() == 3 && path[0] == XmlTagName::k_EnumerationResults
```
Comment (katmsft) [inline sdk/storage/inc/blobs/internal/protocol/blob_rest_client.hpp:1211]: I think deeply nested `if` statements are more readable than this approach.

File: sdk/storage/inc/blobs/internal/protocol/blob_rest_client.hpp [hunk 2]
```diff
@@ -1002,6 +896,304 @@ namespace Azure { namespace Storage { namespace Blobs {
         return ListBlobContainersParseResponse(*response);
       }
 
+    private:
+      static ListContainersSegment ListContainersSegmentFromXml(XmlReader& reader)
```
Comment (katmsft) [inline sdk/storage/inc/blobs/internal/protocol/blob_rest_client.hpp:1144]: If parsing XML does not except on partial XML, which sometimes gets returned by service due to unstable internet connection, then we should track if the content-length is being checked against the actual length of the response body within the Http Stack itself. Can you please create an item to track this?

File: sdk/storage/inc/blobs/internal/protocol/blob_rest_client.hpp (hunk 1 above)
Comment (JinmingHu-MSFT) [reply sdk/storage/inc/blobs/internal/protocol/blob_rest_client.hpp:1211]: but it's harder for auto-generator to generate. It's not a big deal right?

File: sdk/storage/inc/blobs/internal/protocol/blob_rest_client.hpp (hunk 2 above)
Comment (JinmingHu-MSFT) [reply sdk/storage/inc/blobs/internal/protocol/blob_rest_client.hpp:1144]: From the rest client's point of view, the HTTP client returns a incomplete HTTP body. In this case HTTP client should throw an exception, right?

File: sdk/storage/CMakeLists.txt [hunk 3]
```diff
@@ -18,11 +18,11 @@ set (AZURE_STORAGE_BLOB_HEADER
     inc/common/crypt.hpp
     inc/common/xml_wrapper.hpp
     inc/blobs/blob.hpp
+    inc/blobs/blob_service_client.hpp
+    inc/blobs/blob_container_client.hpp
     inc/blobs/blob_client.hpp
     inc/blobs/block_blob_client.hpp
-    inc/blobs/blob_container_client.hpp
     inc/blobs/blob_client_options.hpp
```
Comment (katmsft) [inline sdk/storage/CMakeLists.txt:27]: `blob_client_options` should now be renamed to `blob_options`, given it contains client options and request options for the entire blob service.

File: sdk/storage/inc/blobs/internal/protocol/blob_rest_client.hpp (hunk 1 above)
Comment (JinmingHu-MSFT) [reply sdk/storage/inc/blobs/internal/protocol/blob_rest_client.hpp:1211]: it's more about code style

File: sdk/storage/inc/blobs/block_blob_client.hpp [hunk 4]
```diff
@@ -39,24 +39,30 @@ namespace Azure { namespace Storage { namespace Blobs {
         const std::string& blobUri,
         const BlockBlobClientOptions& options = BlockBlobClientOptions());
 
-    BlockBlobClient WithSnapshot(const std::string& snapshot);
+    BlockBlobClient WithSnapshot(const std::string& snapshot) const;
```
Comment (katmsft) [inline sdk/storage/inc/blobs/block_blob_client.hpp:42]: `WithSnapshot` sounds like a member function that modifies the member instead of creating a new client. This is aligned with .NET but I still think .NET is not having a good name here.
`CloneWithSnapshot` maybe?

Comment (JinmingHu-MSFT) [reply sdk/storage/inc/blobs/block_blob_client.hpp:42]: I prefer `WithSnapshot` since it's aligned with .Net and Js

File: sdk/storage/inc/blobs/blob_service_client.hpp [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+    // anonymous/SAS/customized pipeline auth
+    explicit BlobServiceClient(
+        const std::string& serviceUri,
+        const BlobServiceClientOptions& options = BlobServiceClientOptions());
+
+    ListContainersSegment ListBlobContainersSegment(
+        const ListBlobContainersOptions& options = ListBlobContainersOptions()) const;
+
+    UserDelegationKey GetUserDelegationKey(
+        const std::string& expiresOn,
```
Comment (katmsft) [inline sdk/storage/inc/blobs/blob_service_client.hpp:44]: nit: What is the indentation guide here. I'm seeing 2/4 spaces mixed.

File: sdk/storage/CMakeLists.txt (hunk 3 above)
Comment (JinmingHu-MSFT) [reply sdk/storage/CMakeLists.txt:27]: okay

File: sdk/storage/inc/blobs/blob_service_client.hpp (hunk 5 above)
Comment (JinmingHu-MSFT) [reply sdk/storage/inc/blobs/blob_service_client.hpp:44]: i don't know. it's formatted by clang-format, and I think it's okay.

File: sdk/storage/inc/blobs/blob_service_client.hpp [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+
+    // anonymous/SAS/customized pipeline auth
+    explicit BlobServiceClient(
+        const std::string& serviceUri,
+        const BlobServiceClientOptions& options = BlobServiceClientOptions());
+
+    ListContainersSegment ListBlobContainersSegment(
+        const ListBlobContainersOptions& options = ListBlobContainersOptions()) const;
+
+    UserDelegationKey GetUserDelegationKey(
```
Comment (katmsft) [inline sdk/storage/inc/blobs/blob_service_client.hpp:43]: Why would UserDelegationKey contain below members? They don't seem to be related to UserDelegationKey.
```
    std::string RequestId;
    std::string Date;
    std::string Version;
    std::string ClientRequestId;
```

Comment (JinmingHu-MSFT) [reply sdk/storage/inc/blobs/blob_service_client.hpp:43]: Because we don't have `Response<T>` yet, these members should be put in `Response<T>`.

File: sdk/storage/inc/blobs/internal/protocol/blob_rest_client.hpp (hunk 2 above)
Comment (katmsft) [reply sdk/storage/inc/blobs/internal/protocol/blob_rest_client.hpp:1144]: Yes, but CPPRestsdk does not, so we have this difficult to detect bug. It is due to a previous limitation of the Http stack that we should make sure don't exist in track 2.

Comment (JinmingHu-MSFT) [reply sdk/storage/inc/blobs/internal/protocol/blob_rest_client.hpp:1144]: I guess we can add a check, if request is not `HEAD` and length of HTTP response body doesn't match `Content-Length`, we throw an exception.

File: sdk/storage/inc/blobs/blob_service_client.hpp (hunk 6 above)
Comment (katmsft) [reply sdk/storage/inc/blobs/blob_service_client.hpp:43]: I don't agree with this. I think the Response<T> will only have two things:
1. `GetResponse()`: This API gets the raw response from the request.
2. `GetValue()`: This API gets the `<T>` value parsed by Storage SDK's protocol layer.
We cannot force `Response<T>` to have these four members since the type is shared across service while the four response headers are only used in storage.

Comment (JinmingHu-MSFT) [reply sdk/storage/inc/blobs/blob_service_client.hpp:43]: once `Response<T>` is ready, i'll remove these four fields from every API response. But now, if I removed it, the user had no way to check the `client-request-id`.

File: sdk/storage/inc/blobs/internal/protocol/blob_rest_client.hpp (hunk 2 above)
Comment (JinmingHu-MSFT) [reply sdk/storage/inc/blobs/internal/protocol/blob_rest_client.hpp:1148]: fixed

File: sdk/storage/inc/blobs/blob_service_client.hpp (hunk 6 above)
Comment (katmsft) [reply sdk/storage/inc/blobs/blob_service_client.hpp:43]: I don't think this should be removed. Customers should be able to easily get those values without looking into the raw response.
This can be discussed in the future.
Also, please consider the E2E scenario for this. Customers should be able to use a UserDelegationKey to generate a UserDelegationSas, which makes this return value an input value for some of our API, which will be more bizarre.

Comment (JinmingHu-MSFT) [reply sdk/storage/inc/blobs/blob_service_client.hpp:43]: @katmsft We can add these APIs in the future

File: sdk/storage/inc/blobs/internal/protocol/blob_rest_client.hpp (hunk 2 above)
Comment (katmsft) [reply sdk/storage/inc/blobs/internal/protocol/blob_rest_client.hpp:1148]: Does this checking work when a stream is used for response body?
BTW I don't think it should be here, I think we should ask Http stack to have this check.

File: sdk/storage/inc/blobs/blob_service_client.hpp (hunk 6 above)
Comment (JinmingHu-MSFT) [reply sdk/storage/inc/blobs/blob_service_client.hpp:43]: I created an issue https://github.com/Azure/azure-sdk-for-cpp/issues/176 to track this

File: sdk/storage/inc/blobs/internal/protocol/blob_rest_client.hpp (hunk 2 above)
Comment (JinmingHu-MSFT) [reply sdk/storage/inc/blobs/internal/protocol/blob_rest_client.hpp:1130]: removed

**Comment by vinjiang** [comment]
@gilbertw , this PR seems does not involve your team for review by default, it could because the changes happened to storage folder only. Do you want to add your team to the default reviewer even the change is storage files only?

Meanwhile, we will keep this PR open for one more day in casa any comments you want to put.

**Comment by gilbertw** [comment]
@vinjiang It doesn't include us because of code ownership.  I've added my team to the Storage folders code owner so that we get notified for future PRs.  Thank you for bringing this to my attention
